### PR TITLE
bpo-42532: Check if NonCallableMock's spec_arg is not None instead of call its __bool__ function

### PR DESCRIPTION
**Pull request by idanw206**
https://bugs.python.org/issue42532

If an object has logic in its `__bool__` function, it'll be called here, but it shouldn't be called.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@idanw206

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Misc/NEWS.d/next/Library/2020-12-02-07-37-59.bpo-42532.ObNep_.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Check if NonCallableMock.__new__'s spec_arg is not None instead of calling its __bool__ function
```
Comment (mariocj89) [inline Misc/NEWS.d/next/Library/2020-12-02-07-37-59.bpo-42532.ObNep_.rst]: Suggestion: I'd rephrase this to make it user-oriented. The user probably does not care if we compare to none, they would just see that we were unexpectedly calling the function.

```suggestion
Remove unexpected call of ``__bool__`` when passing a ``spec_arg`` argument to a Mock.
```

**Review by mariocj89 — COMMENTED** [review]
Nicely spotted bug!

**Comment by idanw206** [comment]
> Nicely spotted bug!

Thanks! :)

**Review by mariocj89 — APPROVED** [review]

**Review by cjw296 — CHANGES_REQUESTED** [review]
Nice spot, but please can you add a test that fails before your fix and passes after it.

**Comment by idanw206** [comment]
I have made the requested changes; please review again.

**Review by cjw296 — APPROVED** [review]

**Comment by cjw296** [comment]
@tirkarthi - I've taken off the backport labels since neither of the ones I added appear to have worked. Where should this get backported and how do we do that nowadays?

**Comment by tirkarthi** [comment]
There is a tool called cherry_picker that needs to be installed and can be run locally to cherry pick the commits. It seems bot cannot do it automatically due to conflicts. I will raise manual backports.

Dev guide : https://devguide.python.org/committing/?highlight=Backport#backporting-changes-to-an-older-version

**Comment by cjw296** [comment]
I guess I'm concerned that there are conflicts, it doesn't feel like there should be?

**Comment by tirkarthi** [comment]
Yes, it might affect mock backport. The change is simple enough so not sure if it's in mock.py or the tests. I will update here once I try the backport to 3.9 and 3.8.

**Comment by tirkarthi** [comment]
@cjw296 I get below patch as conflict while trying to cherry pick. I guess 3.8 doesn't need this since it seems to iterate through all args and check for spec which has been simplified in 3.9 and above with #16099. The simplification has a backport to 3.8 in https://github.com/python/cpython/pull/16137 but was not applied. Hence backporting this PR would need #16137 also to be applied. 

```patch
diff --cc Lib/unittest/mock.py
index 3629cf6109,4db1bacf4b..0000000000
--- a/Lib/unittest/mock.py
+++ b/Lib/unittest/mock.py
@@@ -403,18 -402,12 +403,25 @@@ class NonCallableMock(Base)
          # so we can create magic methods on the
          # class without stomping on other mocks
          bases = (cls,)
 -        if not issubclass(cls, AsyncMockMixin):
 +        if not issubclass(cls, AsyncMock):
              # Check if spec is an async object or function
++<<<<<<< HEAD
 +            sig = inspect.signature(NonCallableMock.__init__)
 +            bound_args = sig.bind_partial(cls, *args, **kw).arguments
 +            spec_arg = [
 +                arg for arg in bound_args.keys()
 +                if arg.startswith('spec')
 +            ]
 +            if spec_arg:
 +                # what if spec_set is different than spec?
 +                if _is_async_obj(bound_args[spec_arg[0]]):
 +                    bases = (AsyncMockMixin, cls,)
++=======
+             bound_args = _MOCK_SIG.bind_partial(cls, *args, **kw).arguments
+             spec_arg = bound_args.get('spec_set', bound_args.get('spec'))
+             if spec_arg is not None and _is_async_obj(spec_arg):
+                 bases = (AsyncMockMixin, cls)
++>>>>>>> c598a04dd2... bpo-42532: Check if NonCallableMock's spec_arg is not None instead of call its __bool__ function (GH23613)
          new = type(cls.__name__, bases, {'__doc__': cls.__doc__})
          instance = _safe_super(NonCallableMock, cls).__new__(new)
          return instance
```

**Comment by cjw296** [comment]
It's unclear to me why https://github.com/python/cpython/pull/16099 doesn't have any tests as part of the commit. That would help us figure out whether both or neither of these PRs should end up in 3.8...

**Comment by tirkarthi** [comment]
My guess is that it's a refactoring PR where all bound_args are iterated and checking for starting with "spec" might give others starting with spec like "specfoo" which could have been neither supported/tested. The PR probably tries to be explicit about checking for only spec_set and spec.

**Comment by cjw296** [comment]
@idanw206 - this has been released in the backport now: https://pypi.org/project/mock/4.0.3/
